value/structreflect.go

```diff
@@ -0,0 +1,238 @@
+	fieldPath [][]int
+}
+
+func (f *fieldCacheEntry) getFieldFromStruct(structVal reflect.Value) reflect.Value {
+	// field might be nested within 'inline' structs
+	for _, elem := range f.fieldPath {
+		structVal = structVal.FieldByIndex(elem)
+	}
+	return structVal
+}
+
+func getStructCacheEntry(t reflect.Type) structCacheEntry {
+	if hints, ok := reflectStructCache.Get(t); ok {
+		return hints
+	}
+
+	hints := map[string]*fieldCacheEntry{}
+	buildStructCacheEntry(t, hints, nil)
+
+	reflectStructCache.Update(t, hints)
+	return hints
+}
+
+func buildStructCacheEntry(t reflect.Type, infos map[string]*fieldCacheEntry, fieldPath [][]int) {
+	for i := 0; i < t.NumField(); i++ {
+		field := t.Field(i)
+		jsonName, isInline, isOmitempty := lookupJsonTags(field)
+		if isInline {
+			buildStructCacheEntry(field.Type, infos, append(fieldPath, field.Index))
+			continue
+		}
+		info := &fieldCacheEntry{isOmitEmpty: isOmitempty, fieldPath: append(fieldPath, field.Index)}
+		infos[jsonName] = info
+
+	}
+}
+
+type structReflect struct {
+	Value reflect.Value
+}
+
+func (r structReflect) Length() int {
+	i := 0
+	eachStructField(r.Value, func(s string, value reflect.Value) bool {
+		i++
+		return true
+	})
+	return i
+}
+
+func (r structReflect) Get(key string) (Value, bool) {
+	if val, ok := r.findJsonNameField(key); ok {
+		return mustWrapValueReflect(val), true
+	}
+	return nil, false
+}
+
+func (r structReflect) Has(key string) bool {
+	_, ok := r.findJsonNameField(key)
+	return ok
+}
+
+func (r structReflect) Set(key string, val Value) {
+	fieldVal, ok := r.findJsonNameField(key)
+	if !ok {
+		panic(fmt.Sprintf("key %s may not be set on struct %T: field does not exist", key, r.Value.Interface()))
+	}
+	if !fieldVal.CanSet() {
+		// See https://blog.golang.org/laws-of-reflection for details on why a struct may not be settable
+		panic(fmt.Sprintf("key %s may not be set on struct: %T: struct is not settable", key, r.Value.Interface()))
+	}
+	fieldVal.Set(reflect.ValueOf(val.Unstructured()))
+}
+
+func (r structReflect) Delete(key string) {
+	fieldVal, ok := r.findJsonNameField(key)
+	if !ok {
+		panic(fmt.Sprintf("key %s may not be deleted on struct %T: field does not exist", key, r.Value.Interface()))
+	}
+	if !fieldVal.CanSet() {
+		// See https://blog.golang.org/laws-of-reflection for details on why a struct may not be settable
+		panic(fmt.Sprintf("key %s may not be deleted on struct: %T: struct is not settable", key, r.Value.Interface()))
+	}
+	fieldVal.Set(reflect.Zero(fieldVal.Type()))
+}
+
+func (r structReflect) Iterate(fn func(string, Value) bool) bool {
+	return eachStructField(r.Value, func(s string, value reflect.Value) bool {
+		v := mustWrapValueReflect(value)
+		defer v.Recycle()
+		return fn(s, v)
+	})
+}
+
+func eachStructField(structVal reflect.Value, fn func(string, reflect.Value) bool) bool {
+	for jsonName, fieldCacheEntry := range getStructCacheEntry(structVal.Type()) {
+		fieldVal := fieldCacheEntry.getFieldFromStruct(structVal)
```

Comment:
Will this skip fields with `json:"-"`? I think it should?

Reply:
Fixed and added a test.